### PR TITLE
Inconisistency on whether q_sqrt is triangular fixed.

### PR DESCRIPTION
**Pull request by condnsdmatters**
Problem:
* KL divergence calculations only take lower triangle parts of q_sqrt in
their calculation (implying a lower triangular parameterisation of
covariance)
* Conditional calculations perform calculations using the whole of
q_sqrt.
* The end result is if you forgot to add the triangular transform
to the q_sqrt, your forward pass may be the same (if you initialise with
a triangular matrix), but your gradients will immediately stop q_sqrt
being triangular.

Fix:
* Conditional calulcations now only take the lower triangle parts of q_sqrt
in their calculation

### CODE REVIEW

File: tests/test_kldiv.py [hunk 1]
```diff
@@ -17,13 +17,40 @@
...
 
 rng = np.random.RandomState(0)
 
+# ------------------------------------------
+# Fixtures
+# ------------------------------------------
+
+Ln = 2
+Nn = 10
+Mn = 50
```
Comment (st--) [inline tests/test_kldiv.py]: Would the test be faster but just as useful if these were smaller ?

File: tests/test_kldiv.py [hunk 2]
```diff
@@ -150,3 +177,37 @@ def test_unknown_size_inputs():
     unknown_shape = gauss_kl(mu, sqrt)
 
     np.testing.assert_allclose(known_shape, unknown_shape)
+
+
+@pytest.mark.parametrize('white', [True, False])
+def test_q_sqrt_constraints(inducing_points, kernel, mu, white):
```
Comment (st--) [inline tests/test_kldiv.py]: Would it be worth unifying the different ways of specifying input arrays (pytest fixtures vs Datum class)? I don't have any preference either way...

File: tests/test_conditionals.py [hunk 3]
```diff
@@ -128,3 +130,40 @@ def test_gaussian_whiten(Xdata, Xnew, kernel, mu, sqrt):
...
+    q_sqrt_constrained = Parameter(tril, transform=triangular())
+    q_sqrt_unconstrained = Parameter(tril)
+
+    diff_before_gradient_step = (q_sqrt_constrained - q_sqrt_unconstrained).numpy()
+    assert_allclose(diff_before_gradient_step, 0)
+
+    Fstars = []
+    for q_sqrt in [q_sqrt_constrained, q_sqrt_unconstrained]:
+
+        with tf.GradientTape() as g:
```
Comment (awav) [inline tests/test_conditionals.py]:
```suggestion
        with tf.GradientTape() as tape:
```

File: tests/test_conditionals.py [hunk 4]
```diff
@@ -128,3 +130,40 @@ def test_gaussian_whiten(Xdata, Xnew, kernel, mu, sqrt):
...
+            _, Fstar_var = conditional(Xnew,
+                                       Xdata,
+                                       kernel,
+                                       mu,
+                                       q_sqrt=q_sqrt,
+                                       white=white)
+
+        grad = g.gradient(Fstar_var, q_sqrt.unconstrained_variable)
```
Comment (awav) [inline tests/test_conditionals.py]:
```suggestion
        grad = tape.gradient(Fstar_var, q_sqrt.unconstrained_variable)
```

File: tests/test_conditionals.py [hunk 5]
```diff
@@ -128,3 +130,40 @@ def test_gaussian_whiten(Xdata, Xnew, kernel, mu, sqrt):
 
     assert_allclose(mean_diff, 0, atol=4)
     assert_allclose(var_diff, 0, atol=4)
+
+@pytest.mark.parametrize('white', [True, False])
+def test_q_sqrt_constraints(Xdata, Xnew, kernel, mu, white):
```
Comment (awav) [inline tests/test_conditionals.py]: What's about the previous code, would this test pass?

File: tests/test_kldiv.py [hunk 6]
```diff
@@ -150,3 +177,37 @@ def test_unknown_size_inputs():
...
+    q_sqrt_constrained = Parameter(tril, transform=triangular())
+    q_sqrt_unconstrained = Parameter(tril)
+
+    diff_before_gradient_step = (q_sqrt_constrained - q_sqrt_unconstrained).numpy()
+    assert_allclose(diff_before_gradient_step, 0)
+
+    kls = []
+    for q_sqrt in [q_sqrt_constrained, q_sqrt_unconstrained]:
+
+        with tf.GradientTape() as g:
```
Comment (awav) [inline tests/test_kldiv.py]:
```suggestion
        with tf.GradientTape() as tape:
```

File: tests/test_kldiv.py [hunk 7]
```diff
@@ -150,3 +177,37 @@ def test_unknown_size_inputs():
...
+            kl = prior_kl(inducing_points, kernel, mu, q_sqrt, whiten=white)
+
+        grad = g.gradient(kl, q_sqrt.unconstrained_variable)
```
Comment (awav) [inline tests/test_kldiv.py]:
```suggestion
        grad = tape.gradient(kl, q_sqrt.unconstrained_variable)
```

**Comment by awav** [comment]
@condnsdmatters, I left some comments for you ;)

File: gpflow/conditionals/util.py [hunk 8]
```diff
@@ -82,7 +82,7 @@ def base_conditional(Kmn: tf.Tensor,
         if q_sqrt_dims == 2:
             LTA = A * tf.expand_dims(tf.transpose(q_sqrt), 2)  # [R, M, N]
         elif q_sqrt_dims == 3:
-            L = q_sqrt
+            L =  tf.linalg.band_part(q_sqrt, -1, 0)  # force lower triangle # [R, M, M]
```
[the remark this answers is not in the document]
Comment (st--) [reply gpflow/conditionals/util.py]: Yes, using proper distributions and passing a q(u) to conditionals is definitely something that we should move towards to :)

File: tests/test_kldiv.py (hunk 1 above)
Comment (st--) [reply tests/test_kldiv.py]: (it may not be much difference for any individual test, but with lots of tests even milliseconds can start to matter)

**Comment by st--** [comment]
@condnsdmatters does this address #846 ?

File: tests/test_conditionals.py (hunk 5 above)
Comment (condnsdmatters) [reply tests/test_conditionals.py]: no

File: tests/test_kldiv.py (hunk 2 above)
Comment (condnsdmatters) [reply tests/test_kldiv.py]: pytests fixtures have the advantage of only running once then making the most of pytests in built caching. In general we should move to fixtures, though I also want to look at different ways you can sort of parameterize fixtures. (ie run with multiple diff fixtures)

**Pull request merged**